OrbitCore/CaptureData.cpp

```diff
@@ -14,6 +14,30 @@
 using orbit_client_protos::FunctionInfo;
 using orbit_client_protos::FunctionStats;
 using orbit_client_protos::LinuxAddressInfo;
+using orbit_client_protos::ThreadStateSliceInfo;
+
+void CaptureData::ForEachThreadStateSliceIntersectingTimeRange(
+    int32_t thread_id, uint64_t min_timestamp, uint64_t max_timestamp,
+    const std::function<void(const ThreadStateSliceInfo&)>& action) const {
+  absl::MutexLock lock{thread_state_slices_mutex_.get()};
```

Comment:
I wonder if contention is going to become an issue when drawing as we receive timers... we'd need to profile eventually.

Reply:
Locking when not capturing is not a problem, I was talking specifically about contention when we are capturing and rendering at the same time. It seems the rendering thread could stall the receiving thread for the duration of the thread state track render. The blockchain data structure was made partly for exactly this problem. The render thread would start by querying the current number of timers, and then it could iterate without locking because the data is never migrated (as opposed to a std::vector resize operation for example).

Reply:
We should just rethink this, and then apply it to the whole `CaptureData`. I know about `BlockChain`, but for example it then doesn't allow binary search&mdash;in the worst case of being completely zoomed out this doesn't make a difference, but still. As you know `TimerChain` solves this by having a min timestamp and max timestamp for each block, which is already much better, but it's only for `TextBox`es, and in my opinion exposes the internal block structure too much for use everywhere. If we want to go this route, we could generalize it and polish it to get a generic lock-less single-producer single-consumer data structure that supports somewhat-efficient searching based on time, but it requires more planning.